src/app.ts

```diff
@@ -1,26 +1,65 @@
-// example interfaces that can be use
-// TIP: the types mentioned in the interfaces must be fulfilled in order to solve the problem.
-interface TemperatureReading {
-  time: Date
-  temperature: number
-  city: string
-}
-interface TemperatureSummary {
-  first: number
-  last: number
-  high: number
-  low: number
-  average: number
+import { TemperatureReading, TemperatureSummary } from './interfaces'
+
+type ProccesedReadings = {
+  [city: string]: {
+    [date: string]: number[]
+  }
 }
+let processedReadings: ProccesedReadings = {}
 
 export function processReadings(readings: TemperatureReading[]): void {
-  // add here your code
+  const processedData = readings.reduce<ProccesedReadings>((data, value) => {
+    const { city, time, temperature } = value
+    const formatedDate = time.toISOString()
+
+    // Validate if city exists or isn't empty
+    if (city.trim() === '' || !city) {
+      throw new Error('Some register has no valid city')
+    }
+
+    // Create city objects if not exists
+    if (!data[city]) {
+      data[city] = {}
+    }
+
+    // Create date objects if not exists
+    if (!data[city][formatedDate]) {
+      data[city][formatedDate] = []
+    }
+
+    // Push temperatures
+    data[city][formatedDate].push(temperature)
+    return data
+  }, {} as ProccesedReadings)
+
+  processedReadings = { ...processedData }
 }
 
 export function getTemperatureSummary(
   date: Date,
   city: string,
 ): TemperatureSummary | null {
-  //add here your code
-  return null
+  const formatedDate = date.toISOString()
+
+  if (!processedReadings[city] || !processedReadings[city][formatedDate])
+    return null
+
+  const temperatures = processedReadings[city][formatedDate]
+
+  const first = temperatures[0]
+  const last = temperatures[temperatures.length - 1]
+  const high = Math.max(...temperatures)
+  const low = Math.min(...temperatures)
+  const average =
+    temperatures.reduce((acc, value) => acc + value) / temperatures.length
+
+  const summary: TemperatureSummary = {
+    first,
+    last,
+    high,
+    low,
+    average,
+  }
+
+  return summary
```

Comment:
You can return an object instead of creating an additional object and adding a type. The type is not needed since the function is expecting to return either TemperatureSummary or null.

Reply:
You are correct, type summary constant is not necessary because I'm already typing the return of the function, so, this is redundant. 

I created summary object to make my code more visually understandable, calling this constant as summary. but I suppose some devs can infer this by return type of my function. When would you recommend being explicit with my variables to make my code more understandable?